cppcache/src/PdxHelper.cpp

```diff
@@ -47,92 +47,105 @@ PdxHelper::PdxHelper() {}
 
 PdxHelper::~PdxHelper() {}
 
-void PdxHelper::serializePdx(
-    DataOutput& output, const std::shared_ptr<PdxSerializable>& pdxObject) {
-  auto pdxII = std::dynamic_pointer_cast<PdxInstanceImpl>(pdxObject);
-  auto cacheImpl = CacheRegionHelper::getCacheImpl(output.getCache());
-  auto pdxTypeRegistry = cacheImpl->getPdxTypeRegistry();
-  auto& cachePerfStats = cacheImpl->getCachePerfStats();
+std::shared_ptr<PdxInstanceImpl> PdxHelper::getPdxInstance(
+    const std::shared_ptr<PdxSerializable>& pdxObject) {
+  return std::dynamic_pointer_cast<PdxInstanceImpl>(pdxObject);
+}
 
-  if (pdxII != nullptr) {
-    auto piPt = pdxII->getPdxType();
-    if (piPt != nullptr &&
-        piPt->getTypeId() ==
-            0)  // from pdxInstance factory need to get typeid from server
-    {
-      int typeId = pdxTypeRegistry->getPDXIdForType(
-          piPt, DataOutputInternal::getPool(output));
-      pdxII->setPdxId(typeId);
-    }
-    auto plw = PdxLocalWriter(output, piPt, pdxTypeRegistry);
-    pdxII->toData(plw);
-    plw.endObjectWriting();  // now write typeid
-    int len = 0;
-    uint8_t* pdxStream = plw.getPdxStream(len);
-    pdxII->updatePdxStream(pdxStream, len);
+bool PdxHelper::isPdxInstance(
+    const std::shared_ptr<PdxSerializable>& pdxObject) {
+  return (getPdxInstance(pdxObject) != nullptr);
+}
 
-    delete[] pdxStream;
+bool PdxHelper::hasPdxTypeId(
+    const std::shared_ptr<PdxInstanceImpl>& pdxObject) {
+  auto objectType = pdxObject->getPdxType();
+  return (objectType && (objectType->getTypeId() != 0));
+}
 
-    return;
+void PdxHelper::serializePdxInstance(
+    const std::shared_ptr<PdxInstanceImpl>& pdxInstance,
+    std::shared_ptr<PdxTypeRegistry>& pdxTypeRegistry, DataOutput& output) {
+  if (!hasPdxTypeId(pdxInstance)) {
+    int typeId = pdxTypeRegistry->getPDXIdForType(
+        pdxInstance->getPdxType(), DataOutputInternal::getPool(output));
+    pdxInstance->setPdxId(typeId);
   }
+  auto pdxLocalWriter =
+      PdxLocalWriter(output, pdxInstance->getPdxType(), pdxTypeRegistry);
+  pdxInstance->toData(pdxLocalWriter);
+  pdxLocalWriter.endObjectWriting();  // now write typeid
+  auto pdxStream = pdxLocalWriter.getPdxStream();
+  pdxInstance->updatePdxStream(pdxStream, pdxLocalWriter.getBytesWritten());
+  delete[] pdxStream;
+}
 
+PdxRemoteWriter PdxHelper::createPdxRemoteWriter(
+    const std::shared_ptr<PdxTypeRegistry>& pdxTypeRegistry,
+    const std::shared_ptr<PdxSerializable>& pdxObject, DataOutput& output,
+    const std::string& className) {
+  if (auto pd = pdxTypeRegistry->getPreserveData(pdxObject)) {
+    auto mergedPdxType = pdxTypeRegistry->getPdxType(pd->getMergedTypeId());
+    return PdxRemoteWriter(output, mergedPdxType, pd, pdxTypeRegistry);
+  } else {
+    return PdxRemoteWriter(output, className, pdxTypeRegistry);
+  }
+}
+
+void PdxHelper::serializePdxSerializable(
```

Comment:
Ditch ownership semantics, pass by reference.

Why is cacheImpl passed by pointer? Is it optional? Are you going to check for null? Don't even pass it. Why? Because the code doesn't even use cacheImpl directly, it's a transient dependency - the code depends on cachePerfStats and PdxTypeRegistry (inside regiserPdxType, which needs an overload for the instance you already have a reference to the registry, like we do here).